### DISCUSSION TITLE
Facing issues while using bulk user upload feature in Sunbird ED Portal

### DISCUSSION THREAD

**Discussion by sanjeev-scm** [Sunbird-Lern/Community#160] [Issues] [answered]
Hi @reshmi-nair,

Sunbird release version (tags) - release-4.10.0_RC12

We have tried using bulk user upload for the following API's but getting 400 series errors.
And we tried whitelisting and testing on both the API's
{{host}}/learner/user/v1/upload
{{host}}/learner/user/v2/upload

We would like to know which is the correct API to use for BULK USER UPLOAD.
I have also attached the sample csv file.
Please clarify  if the structure of the csv file is proper.

Please find the details below:

For {{host}}/learner/user/v1/upload :

We are getting 400 - BAD REQUEST ERROR

For {{host}}/learner/user/v2/upload :
The code in release-4.10.0_RC12 points to user/v2/upload API

We are getting 404 Not Found
and the HTML response as 
<!DOCTYPE html>
<html lang="en">
    <head>
        <title>Not Found</title>
        
<style  type="text/css">
            html, body, pre {
                margin: 0;
                padding: 0;
                font-family: Monaco, 'Lucida Console', monospace;
                background: #ECECEC;
            }
            h1 {
                margin: 0;
                background: #AD632A;
                padding: 20px 45px;
                color: #fff;
                text-shadow: 1px 1px 1px rgba(0,0,0,.3);
                border-bottom: 1px solid #9F5805;
                font-size: 28px;
            }
            p#detail {
                margin: 0;
                padding: 15px 45px;
                background: #F6A960;
                border-top: 4px solid #D29052;
                color: #733512;
                text-shadow: 1px 1px 1px rgba(255,255,255,.3);
                font-size: 14px;
                border-bottom: 1px solid #BA7F5B;
            }
        </style>
    </head>
    <body>
        <h1>Not Found</h1>

        <p id="detail">
            For request 'POST /user/v2/upload'
        </p>

    </body>
</html>
[Sample_Users_Bulk Upload.csv](https://github.com/Sunbird-Lern/Community/files/12701020/Sample_Users_Bulk.Upload.csv)

**Comment by AmiableAnil** [comment]
@sanjeev-scm 

- For {{host}}/learner/user/v1/upload :
        Please share the logs for this API.
- For {{host}}/learner/user/v2/upload :
        We have only v1 version of the user upload API in userorg service. [routes](https://github.com/Sunbird-Lern/userorg-service/blob/3d052efac894b4079376b213da55242174602364/controller/conf/routes#L88)

**Reply by Namratakushwah** [reply]
@AmiableAnil Please fine the user upload sample 
[working user bulk upload.csv](https://github.com/Sunbird-Lern/Community/files/12723337/working.user.bulk.upload.csv)

**Reply by sanjeev-scm** [reply]
Hi Anil,


Sunbird release version (tags) - release-4.10.0_RC12

We have used the the same file attached by you. Just changed the orgExternalId
We are getting Bad Request. [Unexpected end of input]

Please refer the attached file.
[working.user.bulk.upload.csv](https://github.com/Sunbird-Lern/Community/files/12727959/working.user.bulk.upload.csv)

Request you to suggest us proceeding further.

Request: 
Request URL: {{host}}/learner/user/v1/upload
Request Method: POST
Status Code: 400 Bad Request

Response:

<!DOCTYPE html>
<html lang="en">
    <head>
        <title>Bad Request</title>
        
<style  type="text/css">
            html, body, pre {
                margin: 0;
                padding: 0;
                font-family: Monaco, 'Lucida Console', monospace;
                background: #ECECEC;
            }
            h1 {
                margin: 0;
                background: #AD632A;
                padding: 20px 45px;
                color: #fff;
                text-shadow: 1px 1px 1px rgba(0,0,0,.3);
                border-bottom: 1px solid #9F5805;
                font-size: 28px;
            }
            p#detail {
                margin: 0;
                padding: 15px 45px;
                background: #F6A960;
                border-top: 4px solid #D29052;
                color: #733512;
                text-shadow: 1px 1px 1px rgba(255,255,255,.3);
                font-size: 14px;
                border-bottom: 1px solid #BA7F5B;
            }
        </style>
    </head>
    <body>
        <h1>Bad Request</h1>

        <p id="detail">
            For request 'POST /v1/user/upload' [Unexpected end of input]
        </p>

    </body>
</html>
cc:  @coolbung @komalm @Renuka-S @vinod303 @sravani-64

**Reply by Jayaprakash8887** [reply]
@sanjeev-scm Please share the logs from user-org service.

**Comment by Nithin-ns24** [comment]
> @sanjeev-scm Please share the logs from user-org service.
@Jayaprakash8887 
Hii we didnt get any logs from user-org service, as we got logs from player and API manager attaching those logs can please you look into this

**API manager logs** 

49.207.244.140 - - [28/Sep/2023:09:35:15 +0000] "POST /user/v1/upload HTTP/1.1" 400 4498 1178 0.089 0.054 . "https://domain-name/dashBoard/addusess" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Safari/537.36" "8d0d9dab-41b8-9f62-eff1-afc3700b1608" "e92746fd34524e5bf8bccb08c02b2516" "0130152125009674241" "dev.nulp.portal" "-" "y9cgaiVJVGymGcYTybIsfNNf0zzL_jKT"


**Player logs**

{"channel":"sunbird","env":"dev.nulp.portal","level":"info","message":{"matchPattern":"/learner/user/v1/upload","method":"POST","msg":"Portal_API_WHITELIST : Middleware for URL [ /learner/user/v1/upload ]","reqPath":"/learner/user/v1/upload","rulesForURL":["ADMIN","ORG_ADMIN"],"sessionRoles":["CONTENT_CREATOR","CONTENT_REVIEWER","COURSE_MENTOR","ORG_ADMIN","PUBLIC","ANONYMOUS"]},"pdata":{"id":"dev.nulp.portal","pid":"sunbird-portal-backend","ver":"4.10.0"},"timestamp":"2023-09-28 09:35:15"}
{
  eid: 'LOG',
  ets: 1695893715810,
  ver: '3.0',
  mid: 'LOG:410215efe3e9e8ca1670373898ea50b8',
  actor: { id: '590a8d0e-de92-4b3f-a9b8-b534e271c54b', type: 'User' },
  context: {
    channel: '0130152125009674241',
    pdata: {
      id: 'dev.nulp.portal',
      ver: '4.10.0',
      pid: 'sunbird-portal-backend'
    },
    env: 'learner-service',
    sid: 'y9cgaiVJVGymGcYTybIsfNNf0zzL_jKT',
    did: 'e92746fd34524e5bf8bccb08c02b2516',
    cdata: [],
    rollup: {
      l1: '0130152125009674241',
      l2: '0130152125009674241',
      l3: '0130152125009674241'
    }
  },
  object: {},
  tags: [ '0130152125009674241' ],
  edata: {
    type: 'api_access',
    level: 'INFO',
    message: '',
    params: [ [Object], [Object], [Object], [Object] ]
  }
}
{"channel":"sunbird","env":"dev.nulp.portal","level":"info","message":{"msg":"/learner/* called - POST - /learner/user/v1/upload"},"pdata":{"id":"dev.nulp.portal","pid":"sunbird-portal-backend","ver":"4.10.0"},"timestamp":"2023-09-28 09:35:15"}
{"channel":"sunbird","env":"dev.nulp.portal","level":"error","message":{"error":"{\"message\":\"Unexpected token < in JSON at position 0\",\"stack\":\"SyntaxError: Unexpected token < in JSON at position 0\\n    at JSON.parse (<anonymous>)\\n    at userResDecorator (/home/sunbird/app_dist/routes/learnerRoutes.js:109:29)\\n    at decorateProxyResBody (/home/sunbird/app_dist/node_modules/express-http-proxy/app/steps/decorateUserRes.js:54:14)\\n    at propagateAslWrapper (/home/sunbird/app_dist/node_modules/async-listener/index.js:504:23)\\n    at /home/sunbird/app_dist/node_modules/async-listener/glue.js:188:31\\n    at proxyWrapper (/home/sunbird/app_dist/node_modules/async-listener/index.js:513:29)\\n    at /home/sunbird/app_dist/node_modules/async-listener/index.js:541:70\\n    at /home/sunbird/app_dist/node_modules/async-listener/glue.js:188:31\\n    at runMicrotasks (<anonymous>)\\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)\"}","msg":"learner route : userResDecorator json parse error:","proxyResData":{"data":[60,33,68,79,67,84,89,80,69,32,104,116,109,108,62,10,60,104,116,109,108,32,108,97,110,103,61,34,101,110,34,62,10,32,32,32,32,60,104,101,97,100,62,10,32,32,32,32,32,32,32,32,60,116,105,116,108,101,62,66,97,100,32,82,101,113,117,101,115,116,60,47,116,105,116,108,101,62,10,32,32,32,32,32,32,32,32,10,60,115,116,121,108,101,32,32,116,121,112,101,61,34,116,101,120,116,47,99,115,115,34,62,10,32,32,32,32,32,32,32,32,32,32,32,32,104,116,109,108,44,32,98,111,100,121,44,32,112,114,101,32,123,10,32,32,32,32,32,32,32,32,32,32,32,32,32,32,32,32,109,97,114,103,105,110,58,32,48,59,10,32,32,32,32,32,32,32,32,32,32,32,32,32,32,32,32,112,97,100,100,105,110,103,58,32,48,59,10,32,32,32,32,32,32,32,32,32,32,32,32,32,32,32,32,102,111,110,116,45,102,97,109,105,108,121,58,32,77,111,110,97,99,111,44,32,39,76,117,99,105,100,97,32,67,111,110,115,111,108,101,39,44,32,109,111,110,111,115,112,97,99,101,59,10,32,32,32,32,32,32,32,32,32,32,32,32,32,32,32,32,98,97,99,107,103,114,111,117,110,100,58,32,35,69,67,69,67,69,67,59,10,32,32,32,32,32,32,32,32,32,32,32,32,125,10,32,32,32,32,32,32,32,32,32,32,32,32,104,49,32,123,10,32,32,32,32,32,32,32,32,32,32,32,32,32,32,32,32,109,97,114,103,105,110,58,32,48,59,10,32,32,32,32,32,32,32,32,32,32,32,32,32,32,32,32,98,97,99,107,103,114,111,117,110,100,58,32,35,65,68,54,51,50,65,59,10,32,32,32,32,32,32,32,32,32,32,32,32,32,32,32,32,112,97,100,100,105,110,103,58,32,50,48,112,120,32,52,53,112,120,59,10,32,32,32,32,32,32,32,32,32,32,32,32,32,32,32,32,99,111,108,111,114,58,32,35,102,102,102,59,10,32,32,32,32,32,32,32,32,32,32,32,32,32,32,32,32,116,101,120,116,45,115,104,97,100,111,119,58,32,49,112,120,32,49,112,120,32,49,112,120,32,114,103,98,97,40,48,44,48,44,48,44,46,51,41,59,10,32,32,32,32,32,32,32,32,32,32,32,32,32,32,32,32,98,111,114,100,101,114,45,98,111,116,116,111,109,58,32,49,112,120,32,115,111,108,105,100,32,35,57,70,53,56,48,53,59,10,32,32,32,32,32,32,32,32,32,32,32,32,32,32,32,32,102,111,110,116,45,115,105,122,101,58,32,50,56,112,120,59,10,32,32,32,32,32,32,32,32,32,32,32,32,125,10,32,32,32,32,32,32,32,32,32,32,32,32,112,35,100,101,116,97,105,108,32,123,10,32,32,32,32,32,32,32,32,32,32,32,32,32,32,32,32,109,97,114,103,105,110,58,32,48,59,10,32,32,32,32,32,32,32,32,32,32,32,32,32,32,32,32,112,97,100,100,105,110,103,58,32,49,53,112,120,32,52,53,112,120,59,10,32,32,32,32,32,32,32,32,32,32,32,32,32,32,32,32,98,97,99,107,103,114,111,117,110,100,58,32,35,70,54,65,57,54,48,59,10,32,32,32,32,32,32,32,32,32,32,32,32,32,32,32,32,98,111,114,100,101,114,45,116,111,112,58,32,52,112,120,32,115,111,108,105,100,32,35,68,50,57,48,53,50,59,10,32,32,32,32,32,32,32,32,32,32,32,32,32,32,32,32,99,111,108,111,114,58,32,35,55,51,51,53,49,50,59,10,32,32,32,32,32,32,32,32,32,32,32,32,32,32,32,32,116,101,120,116,45,115,104,97,100,111,119,58,32,49,112,120,32,49,112,120,32,49,112,120,32,114,103,98,97,40,50,53,53,44,50,53,53,44,50,53,53,44,46,51,41,59,10,32,32,32,32,32,32,32,32,32,32,32,32,32,32,32,32,102,111,110,116,45,115,105,122,101,58,32,49,52,112,120,59,10,32,32,32,32,32,32,32,32,32,32,32,32,32,32,32,32,98,111,114,100,101,114,45,98,111,116,116,111,109,58,32,49,112,120,32,115,111,108,105,100,32,35,66,65,55,70,53,66,59,10,32,32,32,32,32,32,32,32,32,32,32,32,125,10,32,32,32,32,32,32,32,32,60,47,115,116,121,108,101,62,10,32,32,32,32,60,47,104,101,97,100,62,10,32,32,32,32,60,98,111,100,121,62,10,32,32,32,32,32,32,32,32,60,104,49,62,66,97,100,32,82,101,113,117,101,115,116,60,47,104,49,62,10,10,32,32,32,32,32,32,32,32,60,112,32,105,100,61,34,100,101,116,97,105,108,34,62,10,32,32,32,32,32,32,32,32,32,32,32,32,70,111,114,32,114,101,113,117,101,115,116,32,39,80,79,83,84,32,47,118,49,47,117,115,101,114,47,117,112,108,111,97,100,39,32,91,85,110,101,120,112,101,99,116,101,100,32,101,110,100,32,111,102,32,105,110,112,117,116,93,10,32,32,32,32,32,32,32,32,60,47,112,62,10,10,32,32,32,32,60,47,98,111,100,121,62,10,60,47,104,116,109,108,62,10],"type":"Buffer"}},"pdata":{"id":"dev.nulp.portal","pid":"sunbird-portal-backend","ver":"4.10.0"},"timestamp":"2023-09-28 09:35:15"}

**Reply by AmiableAnil** [reply]
@sanjeev-scm @Nithin-ns24 Can you please share file which you are trying to upload? Or can we have call to debug this issue?

@rajeevsathish any input from the portal side?

**Reply by AmiableAnil** [reply]
@sanjeev-scm @Nithin-ns24 Also please share the learner service log.

**Reply by komalm** [reply]
@AmiableAnil PFA of CSV file
[working.user.bulk.upload (2).csv](https://github.com/Sunbird-Lern/Community/files/12789179/working.user.bulk.upload.2.csv)

**Reply by AmiableAnil** [reply]
@komalm please share the log from learner-service.

**Reply by Nithin-ns24** [reply]
@AmiableAnil we are not getting any logs from learner-service for that API, but we are getting logs in player service can u please check this

{"channel":"sunbird","env":"dev.nulp.portal","level":"info","message":{"matchPattern":"/learner/user/v1/upload","method":"POST","msg":"Portal_API_WHITELIST : Middleware for URL [ /learner/user/v1/upload ]","reqPath":"/learner/user/v1/upload","rulesForURL":["ADMIN","ORG_ADMIN"],"sessionRoles":["CONTENT_CREATOR","CONTENT_REVIEWER","COURSE_MENTOR","ORG_ADMIN","PUBLIC","ANONYMOUS"]},"pdata":{"id":"dev.nulp.portal","pid":"sunbird-portal-backend","ver":"4.10.0"},"timestamp":"2023-10-03 12:20:14"}
{
  eid: 'LOG',
  ets: 1696335614303,
  ver: '3.0',
  mid: 'LOG:d5f27db6c36f9a4888ba1691293b673a',
  actor: { id: '590a8d0e-de92-4b3f-a9b8-b534e271c54b', type: 'User' },
  context: {
    channel: '0130152125009674241',
    pdata: {
      id: 'dev.nulp.portal',
      ver: '4.10.0',
      pid: 'sunbird-portal-backend'
    },
    env: 'learner-service',
    sid: 'DIbnbivOz_-qrI1f0L9u3buld-no5UKj',
    did: 'd990c07a24ddc6bf7181429ce3a2b492',
    cdata: [],
    rollup: {
      l1: '0130152125009674241',
      l2: '0130152125009674241',
      l3: '0130152125009674241'
    }
  },
  object: {},
  tags: [ '0130152125009674241' ],
  edata: {
    type: 'api_access',
    level: 'INFO',
    message: '',
    params: [ [Object], [Object], [Object], [Object] ]
  }
}
{"channel":"sunbird","env":"dev.nulp.portal","level":"info","message":{"msg":"/learner/* called - POST - /learner/user/v1/upload"},"pdata":{"id":"dev.nulp.portal","pid":"sunbird-portal-backend","ver":"4.10.0"},"timestamp":"2023-10-03 12:20:14"}
{"channel":"sunbird","env":"dev.nulp.portal","level":"error","message":{"error":"{\"message\":\"Unexpected token < in JSON at position 0\",\"stack\":\"SyntaxError: Unexpected token < in JSON at position 0\\n    at JSON.parse (<anonymous>)\\n    at userResDecorator (/home/sunbird/app_dist/routes/learnerRoutes.js:109:29)\\n    at decorateProxyResBody (/home/sunbird/app_dist/node_modules/express-http-proxy/app/steps/decorateUserRes.js:54:14)\\n    at propagateAslWrapper (/home/sunbird/app_dist/node_modules/async-listener/index.js:504:23)\\n    at /home/sunbird/app_dist/node_modules/async-listener/glue.js:188:31\\n    at proxyWrapper (/home/sunbird/app_dist/node_modules/async-listener/index.js:513:29)\\n    at /home/sunbird/app_dist/node_modules/async-listener/index.js:541:70\\n    at /home/sunbird/app_dist/node_modules/async-listener/glue.js:188:31\\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)\"}","msg":"learner route : userResDecorator json parse error:","proxyResData

**Reply by AmiableAnil** [reply, marked as answer]
@Nithin-ns24 @komalm Can you schedule a call, add me and @rajeevsathish.

**Reply by mohitga** [reply]
@Nithin-ns24 @komalm, i am marking the thread as answered, Please feel free to open the thread in case you have any query on the same.

**Reply by paritshivani** [reply]
Issue is still open
@Nithin-ns24 , can you update latest logs for it?

@AmiableAnil, @rajeevsathish.

**Reply by Nithin-ns24** [reply]
@paritshivani  @komalm @rajeevsathish @AmiableAnil 
Regarding the bulk user upload issue, from the API end point where it is failing it is pointing to "learner" service and we have checked the "learner" service logs, we are not getting any logs from this service. can we schedule a call and discuss further

**Reply by AmiableAnil** [reply]
@komalm @Nithin-ns24 @paritshivani when last time we had a call, we have concluded something. Can you please summarize it here. There was issue something related to JS on button click. Let's have call if required it again. 
cc: @rajeevsathish

**Reply by paritshivani** [reply]
hello @AmiableAnil ,
When can we connect to check the issue?
Unfortunately the person who worked on the inputs is not with us now. So we don't have latest summery on the same.
@komalm , @Nithin-ns24 , @rajeevsathish

**Reply by AmiableAnil** [reply]
@paritshivani we can have call tomorrow 12:30PM.
